File: src/test/java/ar/edu/unlam/tallerweb1/infrastructure/JdbcRepositoryTestIT.java
package ar.edu.unlam.tallerweb1.infrastructure;

import ar.edu.unlam.tallerweb1.domain.model.User;
import ar.edu.unlam.tallerweb1.infrastructure.utils.DataSource;
import ar.edu.unlam.tallerweb1.infrastructure.utils.HSQLDB4TestDataSource;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

import static ar.edu.unlam.tallerweb1.infrastructure.utils.TestsPersistenceHelper.*;

public abstract class JdbcRepositoryTestIT {

    protected Connection connection;
    protected DataSource dataSource = HSQLDB4TestDataSource.instance();

    @BeforeEach
    public void init() throws SQLException {
        connection = dataSource.getConnection();
    }
    @AfterEach
    public void tearDown() throws SQLException {
        connection.createStatement().execute("delete from payment");
        connection.createStatement().execute("delete from shared_expenses");
        connection.createStatement().execute("delete from user");
        connection.createStatement().execute("delete from friends_group");
        connection.close();
    }

    protected User givenAGroupUser(String user, Long friendsGroupId) throws SQLException {
        createUser(connection, user, friendsGroupId);
        return new User(getUserId(connection, user), user);
    }

    protected Long givenAgroup(String groupName) throws SQLException {
        createGroup(connection, groupName);
        return getFriendsGroupId(connection, groupName);
    }
    protected long givenAnExpent(Long friendsGroupId, String user, Double amount, String detail) throws SQLException {
        createExpent(connection, getUserId(connection, user), friendsGroupId, 120.0, "Cena");
        return getExpentId();
    }

    private long getExpentId() throws SQLException {
        final ResultSet results = connection.createStatement().executeQuery("select id from shared_expenses");
        results.next();
        return results.getLong("id");
    }

    protected void givenAPayment(Long user, double amount, Long expentId) throws SQLException {
        final String sql = "INSERT INTO payment (payer, expent_id, amount) VALUES (?,?,?)";
        final PreparedStatement psUser = connection.prepareStatement(sql);
        psUser.setLong(1, user);
        psUser.setLong(2, expentId);
        psUser.setDouble(3, amount);
        psUser.execute();
    }
}

File: src/test/java/ar/edu/unlam/tallerweb1/infrastructure/JdbcRepositoryTestIT.kt
package ar.edu.unlam.tallerweb1.infrastructure

import ar.edu.unlam.tallerweb1.domain.model.User
import ar.edu.unlam.tallerweb1.infrastructure.utils.DataSource
import ar.edu.unlam.tallerweb1.infrastructure.utils.HSQLDB4TestDataSource.Companion.instance
import ar.edu.unlam.tallerweb1.infrastructure.utils.TestsPersistenceHelper.*
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import java.sql.Connection
import java.sql.SQLException

abstract class JdbcRepositoryTestIT {
    protected lateinit var connection: Connection
    var dataSource: DataSource = instance()

    @BeforeEach
    @Throws(SQLException::class)
    fun init() {
        connection = dataSource.connection
    }

    @AfterEach
    @Throws(SQLException::class)
    fun tearDown() {
        connection.createStatement().execute("delete from payment")
        connection.createStatement().execute("delete from shared_expenses")
        connection.createStatement().execute("delete from user")
        connection.createStatement().execute("delete from friends_group")
        connection.close()
    }

    @Throws(SQLException::class)
    protected fun givenAGroupUser(user: String, friendsGroupId: Long): User {
        createUser(connection, user, friendsGroupId)
        return User(getUserId(connection, user), user)
    }

    @Throws(SQLException::class)
    protected fun givenAgroup(groupName: String): Long {
        createGroup(connection, groupName)
        return getFriendsGroupId(connection, groupName)
    }

    @Throws(SQLException::class)
    protected fun givenAnExpent(friendsGroupId: Long, user: String, amount: Double, detail: String): Long {
        createExpent(
            connection,
            getUserId(connection, user),
            friendsGroupId,
            120.0,
            "Cena"
        )
        return expentId
    }

    @get:Throws(SQLException::class)
    private val expentId: Long
        get() {
            val results = connection.createStatement().executeQuery("select id from shared_expenses")
            results.next()
            return results.getLong("id")
        }

    @Throws(SQLException::class)
    protected fun givenAPayment(user: Long, amount: Double, expentId: Long) {
        val sql = "INSERT INTO payment (payer, expent_id, amount) VALUES (?,?,?)"
        val psUser = connection.prepareStatement(sql)
        psUser.setLong(1, user)
        psUser.setLong(2, expentId)
        psUser.setDouble(3, amount)
        psUser.execute()
    }
}
